src/modules/plus/filter_dynamicgpstext.c

```diff
@@ -0,0 +1,1714 @@
+			mlt_log_error(NULL, "filter_dynamicgps.c memory alloc failed, BYE. realloc(size=%d)", crt_nr_point*sizeof(gps_point));
+			pdata->gps_points_size = 0;
+			return;
+		}
+	}
+
+	//debug
+	/*
+	for (i=0; i<5 && i<pdata->gps_points_size; i++)
+		mlt_log_info(NULL, "filter_dynamicgps.c process_tcx gps_points[%d]: lat:%f lon:%f time:%d ",
+			i, pdata->gps_points[i].lat, pdata->gps_points[i].lon, pdata->gps_points[i].time);
+	for (i=pdata->gps_points_size/2; i<pdata->gps_points_size/2+5 && i<pdata->gps_points_size; i++)
+		mlt_log_info(NULL, "filter_dynamicgps.c process_tcx gps_points[%d]: lat:%f lon:%f time:%d ",
+			i, pdata->gps_points[i].lat, pdata->gps_points[i].lon, pdata->gps_points[i].time);
+	for (i=MAX(pdata->gps_points_size-5,0); i<pdata->gps_points_size; i++)
+		mlt_log_info(NULL, "filter_dynamicgps.c process_tcx gps_points[%d]: lat:%f lon:%f time:%d ",
+			i, pdata->gps_points[i].lat, pdata->gps_points[i].lon, pdata->gps_points[i].time);
+	*/
+}
+
+/** Process a .gpx gps location file, result is put in gps_points[],
+ * ignores points without strictly increasing time (or 0)
+*/
+static void process_gpx(mlt_filter filter, const char* file_content, int bytes_read)
+{
+	private_data* pdata = (private_data*)filter->child;
+
+	// sample point from .GPX file (version 1.0), only lat and lon are mandatory but we also need time for calculations
+	/*
+	<trkpt lat="45.227621666484104" lon="25.16533185322889">
+		<ele>868.3005248873908</ele>
+		<time>2020-07-11T09:03:23.000Z</time>
+		<speed>0.0</speed>
+		<geoidheight>36.1</geoidheight>
+		<src>gps</src>
+		<hdop>0.8</hdop>
+		<vdop>1.3</vdop>
+		<pdop>1.6</pdop>
+	</trkpt>
+	*/
+	gps_point crt_point;
+	char* ptext = file_content;
+	const char* last_address = file_content+bytes_read;
+	int i;
+	
+	//mlt_log_info(NULL, "filter_dynamicgps.c entered process_gpx, ptext=%p, last_address=%p", ptext, last_address);
+	//mlt_log_info(NULL, "filter_dynamicgps.c entered process_gpx, ptext=%.100s", ptext);
+	
+	//ptext will point to the current location in the string where we're processing stuff
+	//we will first look for the start of an entry (<trkpt ) then try to parse all tags <> </> until the end tag </trkpt>
+		
+	//before actual processing takes place, we need to know how many points there are in the file so we can allocate exact gps_points
+	int nr_points = 0, crt_nr_point = 0, skipped_points = 0;
+	while (ptext < last_address)
+	{
+		if (strncmp(ptext, "<trkpt", 6)==0)
+			nr_points++;
+		ptext++;
+	}
+	
+	if (pdata->gps_points)
+		free(pdata->gps_points);
+	pdata->gps_points = (gps_point*) malloc(sizeof(gps_point) * nr_points);
+	if (!pdata->gps_points) {
+		mlt_log_error(NULL, "filter_dynamicgpstext.c can't malloc, BYE! (size=%d)", sizeof(gps_point)*nr_points);
+		return;
+	}
+	pdata->gps_points_size = nr_points;
+	
+	//move pointer back to begining
+	ptext = file_content;
+	//find the start of <trkpt field and jump to it
+	while (ptext<last_address && (ptext = strstr(ptext, "<trkpt ")) != NULL)
```

Comment:
I support Dan's comments about using an xml parser. In addition, it does not look like this parsing makes provisions for Unicode. GPX files support unicode so it would be good to test that.

In general, I would suggest to test with GPX files from as many devices as you can since I am sure they will all have some small differences that could cause problems.

Reply:
I tested with multiple sources (I have 3 devices that can provide gps data), even unicode encoded file (UTF-8) - but without unicode characters - they can only appear in extra tags I don't care about (author, program name, etc). This is one of the advantages of the barebones parser, it silently ignores any fields and characters we don't care about, and also doesn't complain about the file following a particular gpx schema (there's 1.0 and 1.1 with small differences in supported tags), come to think of it, the libxml parser might actually be harder to maintain than this.

Reply:
How do I include/compile libxml2 to the project so I can use it?

Reply:
It is already included. But you should move your filter into the "xml" module so that it is in the same folder as other services that depend on libxml2.